### DISCUSSION TITLE
Header: x-amzn-RateLimit-Limit

### DISCUSSION THREAD

**Discussion by ryankirkland** [saleweaver/python-amazon-sp-api#110] [Q&A]
The Selling Partner API docs [Usage Plans and Rate Limits](https://github.com/amzn/selling-partner-api-docs/blob/main/guides/usage-plans-rate-limits/Usage-Plans-and-Rate-Limits.md) mentions a "x-amzn-RateLimit-Limit" header that would indicate some sort of refresh rate when being throttled.

I've not yet seen this header in any of my calls to Orders, Reports, or Sales. Anyone have any insight in how to best set up querying to most efficiently (and pythonically :) ) navigate throttling?

Any insight would be super appreciated! I'm a bit of a noob here, but want to build out some cool automated reporting for my little seller account.

Thanks y'all!

**Comment by FA5I** [comment]
Had the same question - what I did was use the following library: [https://pypi.org/project/ratelimit/](https://pypi.org/project/ratelimit/)

I personally just used 1/(requests per second) and used that as the rate limit as per the docs.

**Reply by ryankirkland** [reply]
Thanks - I'll give this a shot. I've been playing around with a few different libraries (tenacity, backoff, ratelimiter) and ended up defaulting back to my own logic just using a while loop and a try except clause. I'll try the one you provided and let you know if it works!

What I'm finding is I can do like 20 requests at the rate of 1 per second, then get throttled for about 40 seconds or so. Would be super cool to know how long the actual refresh time is. Other APIs I've used will give you some sort of messaging like (x seconds until tokens refreshed) kinda thing.

**Reply by dehidehidehi** [reply]
The rate limits are documented:
https://github.com/amzn/selling-partner-api-docs/tree/main/references

For feeds, for example: 
https://github.com/amzn/selling-partner-api-docs/blob/main/references/feeds-api/feeds_2020-09-04.md

**Reply by ryankirkland** [reply]
Thanks Kevin. I've seen the rate limits. What I don't see is any sort of refresh time. If I have a rate limit of 1 request per second, I should be able to make 1 request per second, right? However, I can do 1 request per second up to 20 requests, then an inconsistent recharge period (blanking on proper terminology for that) as the 21st request will trigger the throttling. I imagine the 20 requests thing is likely a function of the dynamic rate limiting as I am running a fairly small account.

What I haven't seen documented is the time it takes to refresh your allowed requests after throttling - as it does not appear to be as simple as getting 1 more allowed request per second that passes by. Am I missing that in the docs somewhere? I might be going a bit blind from reading through them so much.

**Reply by FA5I** [reply]
I had the same issue - when I use the rate limits in the docs, and use a `time.sleep(1/(req per second))` to make my program wait, I still hit the quota limit after using up all the burst requests. I couldn't figure out why that was happening tbh.

**Reply by ryankirkland** [reply]
Yeah it's not a crazy big deal - even with throttling I can still fairly quickly pull all historical records. I'm just getting caught in how do I make it as efficient as possible and kinda hooked on watch the time to collect 100 records go down as I tweak. Also unsure of best practice in these scenarios and want to learn as I go so I'm not writing the worst code the world has ever seen.

Thanks for taking the time! It's good to know I'm not alone here.

**Reply by yberezkin** [reply]
+1 on this one. Still can't figure out how to make more requests than bucket size is. The "restore rate" does not seems to work like +1 requests per second. Even put `time.sleep(2)` does not solve the issue.
